pyxem/signals/segments.py

```diff
@@ -0,0 +1,444 @@
+
+from traits.trait_errors import TraitError
+
+
+class LearningSegment:
+    _signal_type = "learning_segment"
+
+    def __init__(self, factors, loadings):
+        # Factors as Signal2D
+        self.factors = factors
+        # Corresponding loadings as Signal2D
+        self.loadings = loadings
+
+    def correlate_learning_segments(self, corr_th_factors=0.4,
+                                    corr_th_loadings=0.4):
+        """Iterates through the factors and loadings and calculates the
+        normalized cross-correlation between all factors and all
+        loadings. Factors and loadings are summed if the correlations of
+        both factors and loadings exceed the given thresholds.
+
+        Parameters
+        ----------
+        corr_th_factors : float
+            Correlation threshold for factors. Must be between -1 and 1.
+            Factors and loadings are summed if both factors and loadings
+            have normalized cross-correlations above corr_th_factors and
+            corr_th_loadings, respectively.
+        corr_th_loadings : int, optional
+            Correlation threshold for loadings. Must be between -1 and 1.
+
+        Returns
+        -------
+        learning_segment : LearningSegment
+            LearningSegment where possibly some factors and loadings
+            have been summed.
+        """
+        # If a mask was used during the decomposition, the factors and/or
+        # loadings will contain nan, which must be converted to numbers prior
+        # to the correlations calculations.
+        factors = self.factors.map(lambda x: np.nan_to_num(x), inplace=False)
+        loadings = self.loadings.map(lambda x: np.nan_to_num(x), inplace=False)
+        factors = factors.copy().data
+        loadings = loadings.copy().data
+        correlated_loadings = np.zeros_like(loadings[:1])
+        correlated_factors = np.zeros_like(factors[:1])
+
+        # For each loading and factor, calculate the normalized
+        # cross-correlation to all other loadings and factors, and define
+        # add_indices for those with a value above corr_th_loadings and
+        # corr_th_factors respectively.
+        while np.shape(loadings)[0] > 0:
+            corr_list_loadings = list(map(
+                lambda x: norm_cross_corr(x, template=loadings[0]), loadings))
+            corr_list_factors = list(map(
+                lambda x: norm_cross_corr(x, template=factors[0]), factors))
+
+            add_indices = np.where(list(map(
+                lambda l, f: (l > corr_th_loadings and f > corr_th_factors),
+                corr_list_loadings, corr_list_factors)))
+
+            correlated_loadings = np.append(
+                correlated_loadings,
+                np.array([np.sum(loadings[add_indices], axis=0)]),
+                axis=0)
+            correlated_factors = np.append(
+                correlated_factors,
+                np.array([np.sum(factors[add_indices], axis=0)]),
+                axis=0)
+
```

Comment:
@pc494 do you think this stuff could be vectorized?

Reply:
If time I'll re-rewrite these for readability (once testing is up and covering) ~ (ie a soft yes)

Reply:
Postponed this to a later date.